src/ATokenCheckpointed.sol

```diff
@@ -0,0 +1,307 @@
+    bytes memory fractionalizedVotes =
+      abi.encodePacked(_forVotesToCast, _againstVotesToCast, _abstainVotesToCast);
+    GOVERNOR.castVoteWithReasonAndParams(
+      proposalId, unusedSupportParam, "crowd-sourced vote", fractionalizedVotes
+    );
+  }
+
+  /// @notice Implements the basic logic to mint a scaled balance token.
+  /// @param caller The address performing the mint
+  /// @param onBehalfOf The address of the user that will receive the scaled tokens
+  /// @param amount The amount of tokens getting minted
+  /// @param index The next liquidity index of the reserve
+  /// @return `true` if the the previous balance of the user was 0
+  function _mintScaledWithCheckpoint(
+    address caller,
+    address onBehalfOf,
+    uint256 amount,
+    uint256 index
+  ) internal returns (bool) {
+    bool _mintScaledReturn = _mintScaled(caller, onBehalfOf, amount, index);
+
+    (uint256 _preMintRawBalance, uint256 _postMintRawBalance) = _checkpointRawBalanceOf(onBehalfOf);
+    totalDepositCheckpoints.push(
+      totalDepositCheckpoints.latest() + (_postMintRawBalance - _preMintRawBalance)
+    );
+
+    return _mintScaledReturn;
+  }
+
+  function _rawBalanceOf(address _user) internal returns (uint256) {
+    return _userState[_user].balance;
+  }
+
+  function _checkpointRawBalanceOf(address _user)
+    internal
+    returns (uint256 _previousBalance, uint256 _currentBalance)
+  {
+    (_previousBalance, _currentBalance) = balanceCheckpoints[_user].push(_rawBalanceOf(_user));
+  }
+
+  // Returns the raw (i.e. unrebased) balanceOf the _user at the _blockNumber.
+  function getPastStoredBalance(address _user, uint256 _blockNumber) public returns (uint256) {
+    // TODO could we use getAtProbablyRecentBlock instead?
+    return balanceCheckpoints[_user].getAtBlock(_blockNumber);
+  }
+
+  function getPastTotalDeposits(uint256 _blockNumber) public returns (uint256) {
+    return totalDepositCheckpoints.getAtBlock(_blockNumber);
+  }
+
+  // forgefmt: disable-start
+  //===========================================================================
+  // BEGIN: Aave overrides
+  //===========================================================================
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaledWithCheckpoint function.
+  ///
+  /// @inheritdoc IAToken
+  function mint(
+    address caller,
+    address onBehalfOf,
+    uint256 amount,
+    uint256 index
+  ) external virtual override onlyPool returns (bool) {
+    return _mintScaledWithCheckpoint(caller, onBehalfOf, amount, index);
+  }
+
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaledWithCheckpoint function.
+  ///
+  /// @inheritdoc IAToken
+  function mintToTreasury(uint256 amount, uint256 index) external override onlyPool {
+    if (amount == 0) {
+      return;
+    }
+    _mintScaledWithCheckpoint(address(POOL), _treasury, amount, index);
+  }
```

Comment:
As I'm reviewing this, I actually think we should *not* be checkpointing this function. It's called by the pool when someone calls [`mintToTreasury`](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/pool/Pool.sol#L446-L448).

But mintToTreasury does not involve depositing any underlying tokens into Aave (even though it results in minting aTokens). So there should be no need to checkpoint anything. Adding checkpoints here would actually mess up our internal accounting.

Furthermore, the pool (which is the address that we would be checkpointing for) is not an entity that can vote

Reply:
Don't fully understand here either, but this code seems to indicate the treasury is earning some income, so perhaps minting to the treasury is just giving the treasury aTokens for the income it has already earned in the pool (fees?). If that's the case, checkpointing *would* be warranted, I believe.

https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/PoolLogic.sol#L107

Reply:
> minting to the treasury is just giving the treasury aTokens for the income it has already earned in the pool (fees?). If that's the case, checkpointing would be warranted, I believe

I think that's right: it's minting aTokens to represent the treasury income, which are fees collected at various points, e.g. [here](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/BridgeLogic.sol#L132) and [here](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/FlashLoanLogic.sol#L238-L240).

Now that I think more about it, it probably does make sense to include this. We're snapshotting aToken balances now, not deposits. And, while the treasury won't have made any deposits, it will have an aToken balance because of this. If at some point Aave decides it wants to participate in governance with the aTokens it has earned as fees, it should be able to. Thanks for correcting me 👍 

I should probably add a test for this functionality though

Reply:
Tested in ba4a946

Reply:
@davidlaprade for posterity can you leave a comment here summarizing what `mintToTreasury` is for, when/who calls it, and whether we need to care about it? The commit linked above was rebased away and the most recent comments in this thread are "I think" style comments, e.g. from reading it I'm not certain if that's correct or outdated

Reply:
Sure!

`mintToTreasury` captures/realizes protocol fees by issuing aTokens to the Aave treasury address.

The function is public, so it can be called by anyone. Ernesto from Aave told us in Telegram that it's called by "somebody in the community I think".

We do need to care about it, for the reason @apbendi stated. ATokens are being issued and the underlying asset could potentially be withdrawn (and used for voting). For consistency, therefore, we should checkpoint when it happens and accord the appropriate voting weight to the Treasury -- even if we think it will never express a voting preference.

We do support this function now, and are checkpointing when it is called, as seen by the test added in ba4a9463270b139cff6009ead572490856a6afc8